src/accountsdb/cache.zig

```diff
@@ -0,0 +1,365 @@
+const std = @import("std");
+const sig = @import("../sig.zig");
+const builtin = @import("builtin");
+const zstd = @import("zstd");
+
+const Account = sig.core.Account;
+const Slot = sig.core.time.Slot;
+const Pubkey = sig.core.pubkey.Pubkey;
+
+const LruCacheCustom = sig.common.lru.LruCacheCustom;
+
+/// Stores read-only in-memory copies of commonly used accounts
+pub const AccountsCache = struct {
+    /// Atomically refcounted account
+    pub const CachedAccount = struct {
+        account: Account,
+        /// represents whether this account has been mutated, and shouldn't progress to newer slots
+        // TODO: when we start mutating accounts, make sure to set this field
+        is_dirty: std.atomic.Value(bool) = .{ .raw = false },
+        ref_count: std.atomic.Value(usize),
+
+        pub fn init(allocator: std.mem.Allocator, account: Account) !CachedAccount {
+            return .{
+                .account = try account.clone(allocator),
+                .ref_count = std.atomic.Value(usize).init(1),
+            };
+        }
+
+        pub fn deinit(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            self.account.deinit(allocator);
+            self.* = undefined;
+        }
+
+        pub fn releaseOrDestroy(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            const current_count = self.ref_count.load(.acquire);
+            if (current_count == 1) {
+                self.deinit(allocator);
+                allocator.destroy(self);
+            } else {
+                _ = self.ref_count.fetchSub(1, .acq_rel);
+            }
```

Comment:
There are multiple races here. I'm not going to list them, but here's how it should look:
```suggestion
            const current_count = self.ref_count.fetchSub(1, .release);
            if (current_count == 1) {
                _ = self.ref_count.load(.acquire);
                self.deinit(allocator);
                allocator.destroy(self);
            }
```